sys/thread.c

```diff
@@ -131,16 +136,18 @@ noreturn void thread_exit(int exitcode) {
   fdtab_release(td->td_fdtable);
 
   mtx_lock(&zombie_threads_mtx);
+  TAILQ_INSERT_TAIL(&zombie_threads, td, td_zombieq);
+  mtx_unlock(&zombie_threads_mtx);
 
   critical_enter();
```

Comment:
I don't think I understand the solution you've linked to. The entire function is marked with a comment:
```
Always called with scheduler locked.
```
which suggests that there is an external mechanism which prevents the scheduler from running while a thread is exiting. The thread mutex gets locked only at the very last moment, which also hints at a different synchronization mechanism besides this mutex.

Also, I do not understand what do you mean by:

> as long as scheduler respects `td_lock` mutex.

Do we want to prevent the scheduler from switching context to threads whose mutex is locked? Should the scheduler lock (or, actually, try-lock) the thread mutex when picking next thread? For this to work well w/o obvious deadlock, we would also have to prevent preemption of threads who own their own thread mutex. I'm under the impression that this could lead to complex problems.

I'd appreciate some detailed description of the idea you want me to implement.